hpx/parallel/spmd_block.hpp

```diff
@@ -0,0 +1,153 @@
+    /// to each image. The constraint for the lambda given to the
+    /// define_spmd_block function is to accept a spmd_block as first parameter.
+    struct spmd_block
+    {
+        explicit spmd_block(std::size_t num_images, std::size_t image_id,
+            hpx::lcos::local::barrier & barrier)
+        : num_images_(num_images), image_id_(image_id), barrier_(barrier)
+        {}
+
+        spmd_block(spmd_block &&) = default;
+        spmd_block(spmd_block const &) = delete;
+
+        spmd_block & operator=(spmd_block &&) = default;
+        spmd_block & operator=(spmd_block const &) = delete;
+
+        std::size_t get_num_images() const
+        {
+            return num_images_;
+        }
+
+        std::size_t this_image() const
+        {
+            return image_id_;
+        }
+
+    // FIXME : If 2 images are scheduled in the same thread (not OS-thread)
+    // -> deadlock
+        void sync_all() const
+        {
+           barrier_.get().wait();
+        }
+
+    private:
+        std::size_t num_images_;
+        std::size_t image_id_;
+        mutable std::reference_wrapper<hpx::lcos::local::barrier> barrier_;
+    };
+
+    namespace detail
+    {
+        template <typename F>
+        struct spmd_block_helper
+        {
+            mutable std::shared_ptr<hpx::lcos::local::barrier> barrier_;
+            typename std::decay<F>::type f_;
+            std::size_t num_images_;
+
+            template <typename ... Ts>
+            void operator()(std::size_t image_id, Ts && ... ts) const
+            {
+                spmd_block block(num_images_, image_id, *barrier_);
+                hpx::util::invoke_r<void>(
+                    f_, block, std::forward<Ts>(ts)...);
+            }
+        };
+    }
+
+    template <typename ExPolicy, typename F, typename ... Args>
+    typename util::detail::algorithm_result<ExPolicy>::type
+    define_spmd_block(ExPolicy && policy,
+        std::size_t num_images, F && f, Args && ... args)
+    {
+        static_assert(
+            parallel::execution::is_execution_policy<ExPolicy>::value,
+            "parallel::execution::is_execution_policy<ExPolicy>::value");
+
+        using ftype = typename std::decay<F>::type;
+
+        using first_type =
+                typename hpx::parallel::v2::detail::extract_first_parameter<
+                            decltype(&ftype::operator())>::type;
+
+        using executor_type =
+            typename hpx::util::decay<ExPolicy>::type::executor_type;
+
+        static_assert(std::is_same<spmd_block, first_type>::value,
+            "define_spmd_block() needs a lambda that " \
+            "has at least a spmd_block as 1st argument");
+
+        std::shared_ptr<hpx::lcos::local::barrier> barrier
+            = std::make_shared<hpx::lcos::local::barrier>(num_images);
+
+        return hpx::parallel::executor_traits<
+                typename std::decay<executor_type>::type
+            >::bulk_execute(
```

Comment:
Opt for synchronous `define_spmd_block()`. Done in 4d109e52d23bce52b3b418811faa88a294dd1953